### PR TITLE
Feature/financial years

### PR DESCRIPTION
**Pull request by BerendWouters**
I'm still in doubt about the domain model of the FinancialYear.

I've added some unit tests nonetheless, so we can still change the implementation.

### CODE REVIEW

File: Persistence/Repositories/AttachmentStorage.cs [hunk 1]
```diff
@@ -50,3 +48,11 @@ public void Delete(string blobUri, CancellationToken cancellationToken)
         throw new NotImplementedException();
     }
 }
+
+public interface IAttachmentStorage
```
Comment (vyruz1986) [inline Persistence/Repositories/AttachmentStorage.cs:52]: This interface should remain in the Domain project. See https://stackoverflow.com/a/72399927 for reasons why

File: Persistence/Repositories/BankAccountRepository.cs [hunk 2]
```diff
@@ -41,3 +40,12 @@ public async Task SaveAsync(CancellationToken ct)
         await _context.SaveChangesAsync(ct);
     }
 }
+
+public interface IBankAccountRepository
```
Comment (vyruz1986) [inline Persistence/Repositories/BankAccountRepository.cs:44]: This interface should remain in the Domain project. See https://stackoverflow.com/a/72399927 for reasons why

File: Persistence/Repositories/FinancialYearRepository.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        _context.FinancialYears.Add(financialYear);
+    }
+
+    public async Task SaveChangesAsync(CancellationToken cancellationToken)
+    {
+        await _context.SaveChangesAsync(cancellationToken);
+    }
+}
+
+public interface IFinancialYearRepository
```
Comment (vyruz1986) [inline Persistence/Repositories/FinancialYearRepository.cs:34]: This interface should be in the Domain project. See https://stackoverflow.com/a/72399927 for reasons why

File: Persistence/Repositories/MemberRepository.cs [hunk 4]
```diff
@@ -45,3 +44,13 @@ public async Task Save(CancellationToken cancellationToken)
     }
 
 }
+
+public interface IMemberRepository
```
Comment (vyruz1986) [inline Persistence/Repositories/MemberRepository.cs:48]: This interface should remain in the Domain project. See https://stackoverflow.com/a/72399927 for reasons why

File: Persistence/HaSpManContext.cs [hunk 5]
```diff
@@ -22,9 +22,16 @@ public HaSpManContext(DbContextOptions<HaSpManContext> options)
     public DbSet<BankAccount> BankAccounts { get; set; } = null!;
     public DbSet<BankAccountsWithTotals> BankAccountsWithTotals { get; set; } = null!;
     public DbSet<Transaction> Transactions { get; set; } = null!;
+    public DbSet<FinancialYear> FinancialYears { get; set; } = null!;
 
     protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
-            => optionsBuilder.UseSqlServer(x => x.MigrationsHistoryTable("__EFMigrationsHistory", Schema.HaSpMan));
+    {
+        if (!optionsBuilder.IsConfigured)
```
Comment (vyruz1986) [inline Persistence/HaSpManContext.cs:29]: Why do we need this?

**Review by vyruz1986 — CHANGES_REQUESTED** [review]

**Review by vyruz1986 — CHANGES_REQUESTED** [review]
I'm also missing an EFCore migration. I'm also wondering whether FInancialYear should become an AggregateRoot whjch owns many Transaction objects. The it is now, they would become 2 aggregate roots with relations between them, which is generally frowned upon in DDD.
A better would be to make FinanalYear the aggregate root, and make the current Transaction aggregate root, a value object of a financial year instead. In this case the TransactionRepository would need to be removed, and each interaction with a transaction would need to take please through the FinancialYear. 
In this an EFCore configuration using `OwnsMany(y => y.Transactions)` would be needed from the FinancialYear config

File: Persistence/HaSpManContext.cs (hunk 5 above)
Comment (BerendWouters) [reply Persistence/HaSpManContext.cs:29]: For unit testing purposes, we can't have the provider already configured.

File: Commands/Handlers/FinancialYear/AddFinancialYear/AddFinancialYearCommand.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+namespace Commands.Handlers.FinancialYear.AddFinancialYear;
+
+public record AddFinancialYearCommand(DateOnly StartDate) : IRequest<Guid>;
+
+
+public class AddFinancialYearCommandValidator : AbstractValidator<AddFinancialYearCommand>
+{
+    public AddFinancialYearCommandValidator()
+    {
+        RuleFor(x => x.StartDate).NotEmpty();
```
[the remark this answers is not in the document]
Comment (BerendWouters) [reply Commands/Handlers/FinancialYear/AddFinancialYear/AddFinancialYearCommand.cs:12]: Ah yes, read over the part of the year being a configurable thing.

File: Commands/Handlers/FinancialYear/AddFinancialYear/AddFinancialYearHandler.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    private readonly HaSpManContext _haSpManContext;
+
+    public AddFinancialYearHandler(IFinancialYearRepository financialYearRepository, HaSpManContext haSpManContext)
+    {
+        _financialYearRepository = financialYearRepository;
+        _haSpManContext = haSpManContext;
+    }
+    public async Task<Guid> Handle(AddFinancialYearCommand request, CancellationToken cancellationToken)
+    {
+        await EnsureStartDateNotInExistingFinancialYear(request, cancellationToken);
```
[the remark this answers is not in the document]
Comment (BerendWouters) [reply Commands/Handlers/FinancialYear/AddFinancialYear/AddFinancialYearHandler.cs:20]: Correct, will update once I've figured out the way to handle configurable start of a new financial year.

**Comment by BerendWouters** [comment]
> I'm also missing an EFCore migration. I'm also wondering whether FInancialYear should become an AggregateRoot whjch owns many Transaction objects. The it is now, they would become 2 aggregate roots with relations between them, which is generally frowned upon in DDD. A better would be to make FinanalYear the aggregate root, and make the current Transaction aggregate root, a value object of a financial year instead. In this case the TransactionRepository would need to be removed, and each interaction with a transaction would need to take please through the FinancialYear. In this an EFCore configuration using `OwnsMany(y => y.Transactions)` would be needed from the FinancialYear config
I was opting for that last path too, but I did want to have this discussed. From a domain point of view, this approach makes to most sense. So: FinancialYear becomes the new aggregate root, Transaction is removed and becomes a value object of that newly created aggregate root.
I do wonder about the following though:
The existing transaction handlers are unaware of the Financial year concept. Am I correct in letting them exist, but instead of interacting with the Transaction (which it can't since we removed the ITransactionRepository), it should first figure out what FinancialYear the transaction belongs to, using the `ReceivedDate` of the transaction? This would imply that we could move transactions from one FinancialYear to another (as longs as they're not closed and so on).
The cool thing about using the FinancialYear as the aggregate root is that we only have to check if the transaction can be edited if the FinancialYear isn't closed, and not the transaction themselves. As a consequence, I don't think we'd need to add a `Locked` property on the Transaction.

The reason I didnt yet create a migration was to prevent too many up and down migrations during development.

**Comment by vyruz1986** [comment]
> > I'm also missing an EFCore migration. I'm also wondering whether FInancialYear should become an AggregateRoot whjch owns many Transaction objects. The it is now, they would become 2 aggregate roots with relations between them, which is generally frowned upon in DDD. A better would be to make FinanalYear the aggregate root, and make the current Transaction aggregate root, a value object of a financial year instead. In this case the TransactionRepository would need to be removed, and each interaction with a transaction would need to take please through the FinancialYear. In this an EFCore configuration using `OwnsMany(y => y.Transactions)` would be needed from the FinancialYear config
> 
> I was opting for that last path too, but I did want to have this discussed. From a domain point of view, this approach makes to most sense. So: FinancialYear becomes the new aggregate root, Transaction is removed and becomes a value object of that newly created aggregate root. I do wonder about the following though: The existing transaction handlers are unaware of the Financial year concept. Am I correct in letting them exist, but instead of interacting with the Transaction (which it can't since we removed the ITransactionRepository), it should first figure out what FinancialYear the transaction belongs to, using the `ReceivedDate` of the transaction? This would imply that we could move transactions from one FinancialYear to another (as longs as they're not closed and so on). The cool thing about using the FinancialYear as the aggregate root is that we only have to check if the transaction can be edited if the FinancialYear isn't closed, and not the transaction themselves. As a consequence, I don't think we'd need to add a `Locked` property on the Transaction.
> 
> The reason I didnt yet create a migration was to prevent too many up and down migrations during development.

Indeed that would impose a breaking change in that transactions could not be accessed by themselves, but would need to be accessed through their FY aggregate root.
Which means that would likely translate back to UI changes as well, as you would first need to select the financial year before being able to see/add/edit transactions...

**Comment by BerendWouters** [comment]
I've updated the PR.

No migrations yet, but I think the domain change has been done.

However, some other issues I'm not sure if the current implementation is the best:
- As you mentioned, handling transactions should be done through the aggregate root. I've removed the TransactionRepository, but allowed to find the FY via a transactionId. I'm not really sure if this goes against any practices. It feels off, but there's a difference between the UI and the actual domain model (which I think _is_ allowed in DDD).
- However, by introducing the FY as the aggregate root and DDD stating you should always start at the aggregate root, we will always track all transactions (and attachments) under one FY. This _might_ introduce a performance penalty, but unsure. It's not quite clear to me to what DDD says about including a certain child item, but it feels like a bad practice.
- Querying. DDD is aimed at managing the state of the object you're working on, but is if I remember correctly quite loose about getting data. For that reason, we still have a DbSet of Transactions, but only for querying purposes. I think this is fine, but we should always adhere the no-tracking configuration, when querying the DbContext. In the past I've used a query facade pattern to make this more strict, but that's just convention imho.

**Comment by vyruz1986** [comment]
> * Querying. DDD is aimed at managing the state of the object you're working on, but is if I remember correctly quite loose about getting data. For that reason, we still have a DbSet of Transactions, but only for querying purposes. I think this is fine, but we should always adhere the no-tracking configuration, when querying the DbContext. In the past I've used a query facade pattern to make this more strict, but that's just convention imho.

I think the DbSet will have to go anyway, because iirc an entity cannot have both it's own DbSet as well as be owned by another entity.
Offcourse you could expose an `IQueryable` from the DbContext which does something like `FinancialYears.SelectMany(y => t>Transactions).AsNoTracking()`, or go all the way as you mentioned and create an `I<Entity>QueryStore` next to `I<Entity>Repository` for query purposes

**Comment by BerendWouters** [comment]
I knew this would happen: owned types don't support hierarchy (https://learn.microsoft.com/en-us/ef/core/modeling/owned-entities#current-shortcomings).

File: Domain/FinancialYearConfiguration.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
+﻿namespace Domain;
+
+public class FinancialYearConfiguration
```
Comment (vyruz1986) [inline Domain/FinancialYearConfiguration.cs:3]: This shouldn't be an aggregate root I think? It should just be configuration which is retrieved wherever needed using the .NET `IOptions` pattern

File: Domain/Transaction.cs [hunk 9]
```diff
@@ -131,6 +131,7 @@ public void AddAttachments(ICollection<TransactionAttachment> attachments)
             Attachments.Add(attachment);
         }
     }
+
```
Comment (vyruz1986) [inline Domain/Transaction.cs:134]: ?

File: Persistence/EntityConfigurations/FinancialYearEntityTypeConfiguration.cs [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+public class FinancialYearEntityTypeConfiguration : IEntityTypeConfiguration<FinancialYear>
+{
+    public void Configure(EntityTypeBuilder<FinancialYear> builder)
+    {
+        builder.Property(x => x.StartDate).IsRequired();
+        builder.Property(x => x.EndDate).IsRequired();
+        builder.Property(x => x.IsClosed).IsRequired();
+        // Owned entity types cannot have inheritance hierarchies https://learn.microsoft.com/en-us/ef/core/modeling/owned-entities#current-shortcomings
+        builder.HasMany(x => x.Transactions);
```
Comment (vyruz1986) [inline Persistence/EntityConfigurations/FinancialYearEntityTypeConfiguration.cs:16]: This should be `OwnsMany()` I think?

File: Persistence/Migrations/20230626190616_AddFinancialYearAsAggregateRoot.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    {
+        /// <inheritdoc />
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.AddColumn<Guid>(
+                name: "FinancialYearId",
+                schema: "HaSpMan",
+                table: "Transactions",
+                type: "uniqueidentifier",
+                nullable: true);
```
Comment (vyruz1986) [inline Persistence/Migrations/20230626190616_AddFinancialYearAsAggregateRoot.cs:19]: This shouldn't be nullable eventually, but of course it must be temporarily nullable for all existing transactions. However the should be some migration SQL code in here to add the column (with nullable: true), update the value for all existing transactions, then make it non-nullable for all new transactions

File: Persistence/EntityConfigurations/TransactionConfigurations.cs [hunk 12]
```diff
@@ -5,6 +5,18 @@
 
 namespace Persistence.EntityConfigurations;
 
+public class TransactionConfiguration : IEntityTypeConfiguration<Transaction>
```
Comment (vyruz1986) [inline Persistence/EntityConfigurations/TransactionConfigurations.cs:8]: This should be an Action in the `.OwnsMany()` I commented above

**Review by vyruz1986 — CHANGES_REQUESTED** [review]

File: Persistence/EntityConfigurations/FinancialYearEntityTypeConfiguration.cs (hunk 10 above)
Comment (BerendWouters) [reply Persistence/EntityConfigurations/FinancialYearEntityTypeConfiguration.cs:16]: Yes, but OwnsMany doesn't support inheritance of collections (yet): https://learn.microsoft.com/en-us/ef/core/modeling/owned-entities#current-shortcomings

File: Persistence/EntityConfigurations/FinancialYearEntityTypeConfiguration.cs [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Persistence.EntityConfigurations;
+
+public class FinancialYearEntityTypeConfiguration : IEntityTypeConfiguration<FinancialYear>
+{
+    public void Configure(EntityTypeBuilder<FinancialYear> builder)
+    {
+        builder.Property(x => x.StartDate).IsRequired();
+        builder.Property(x => x.EndDate).IsRequired();
+        builder.Property(x => x.IsClosed).IsRequired();
```
Comment (vyruz1986) [inline Persistence/EntityConfigurations/FinancialYearEntityTypeConfiguration.cs:14]: These shouldn't be required

File: Persistence/Migrations/20230728184205_TransactionsBelongToFinancialYear.cs [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+        migrationBuilder.AlterColumn<Guid>(
+            "FinancialYearId", 
+            schema:"HaspMan",
+            table: "Transactions",
+            nullable: false);
+        migrationBuilder.CreateIndex(
+            name: "IX_Transactions_FinancialYearId",
+            schema: "HaSpMan",
+            table: "Transactions",
+            column: "FinancialYearId");
```
Comment (vyruz1986) [inline Persistence/Migrations/20230728184205_TransactionsBelongToFinancialYear.cs:27]: This only sets the field to non-nullable in the index, however the table should be altered as well

**Review by vyruz1986 — CHANGES_REQUESTED** [review]

**Review by vyruz1986 — APPROVED** [review]

**Pull request merged**